src/backend/distributed/utils/resource_lock.c

```diff
@@ -1196,3 +1216,164 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
+	}
+
+	List *distributedRelationList = NIL;
+	List *lockDistRelDescendantsList = NIL;
+
+	/*
+	 * this is used to enforce the lock order:
+	 * [...A], [...B]
+	 * Where:
+	 * A: Explicitly locked relations
+	 * B: Relations with a foreign key relationships to relationships in A,
+	 *    which are not locked by the lock command by default
+	 *
+	 * e.g.
+	 *  Let view v with tables t1, t2
+	 *  for LOCK v IN * MODE;
+	 *  A: v
+	 *  B: nil
+	 *  the lock command will lock t1 and t2 so no need to populate B
+	 *
+	 *  Let tables t1 and t2 with t2 referencing t1 with a foreign key
+	 *  for TRUNCATE t1 (we need to acquire locks on t1 and t2)
+	 *  A: t1
+	 *  B: t2
+	 *  t2 will not be locked by lock command if we call LOCK t1
+	 *  so we need to add it to B
+	 */
+	List *referencingRelationIds = NIL;
+
+	bool nowait = (configs & DIST_LOCK_NOWAIT) > 0;
+
+	RangeVar *rangeVar = NULL;
+	foreach_ptr(rangeVar, relationList)
+	{
+		bool lockDescendants = rangeVar->inh;
+		Oid relationId = RangeVarGetRelidExtended(rangeVar, AccessShareLock, nowait ?
+												  RVR_NOWAIT : 0, NULL, NULL);
+
+		/*
+		 * Note that allowing the user to lock shards could lead to
+		 * distributed deadlocks due to shards not being locked when
+		 * a distributed table is locked.
+		 * However, because citus.enable_manual_changes_to_shards
+		 * is a guc which is not visible by default, whoever is using this
+		 * guc will hopefully know what they're doing and avoid such scenarios.
+		 */
+		ErrorIfIllegallyChangingKnownShard(relationId);
+
+		/*
+		 * we want to prevent under privileged users to trigger establishing connections,
+		 * that's why we have this additional check. Otherwise, we'd get the errors as
+		 * soon as we execute the command on any node
+		 */
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		if (get_rel_relkind(relationId) == RELKIND_VIEW)
+		{
+			ObjectAddress viewAddress = { 0 };
+			ObjectAddressSet(viewAddress, RelationRelationId, relationId);
+
+			if (!IsObjectDistributed(&viewAddress))
+			{
+				continue;
+			}
+		}
+		else if (!ShouldSyncTableMetadata(relationId))
+		{
+			continue;
+		}
+
+		distributedRelationList = list_append_unique_oid(distributedRelationList,
+														 relationId);
+		lockDistRelDescendantsList = lappend(lockDistRelDescendantsList,
+											 (void *) (&lockDescendants));
+
+		if ((configs & DIST_LOCK_REFERENCING_TABLES) > 0)
+		{
+			CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
+			Assert(cacheEntry != NULL);
+
+			List *referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+
+			Oid referencingRelationId = InvalidOid;
+			foreach_oid(referencingRelationId, referencingTableList)
+			{
+				if (ShouldSyncTableMetadata(referencingRelationId))
+				{
+					referencingRelationIds = lappend_oid(referencingRelationIds,
+														 referencingRelationId);
+				}
+			}
+		}
+	}
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, referencingRelationIds)
```

Comment:
why don't we immediately add referencingRelationIds to distributedRelationList earlier but defer here?

Reply:
https://github.com/citusdata/citus/blob/ac76db83e8a02e5c00923f24a18b2efa3e2acc91/src/backend/distributed/utils/resource_lock.c#L1243-L1265

Reply:
So, 3 tables `tbl1`, `tbl2`, `tbl3`, where we have `tbl2 <- tbl1`

 When I do:
```SQL
LOCK tbl2, tbl3;
```

I'd expect the order to be `tbl2, tbl1, tbl3`. But, we seem to follow: `tbl2, tbl3, tbl1`.

Is that right? If so, why do you think we should insert `tbl3` in between?

Reply:
What you suggest makes sense. I guess I misinterpreted :locking the referencing tables after" as:
* lock referencing tables after the all the tables that are explicitly specified.

instead of:
* lock referencing tables right after the table they reference.

I will switch the locking order to your suggestion.